src/pudl/output/ferc1.py

```diff
@@ -1337,3 +1339,93 @@ def remove_intra_table_calculated_values(exploded: pd.DataFrame) -> pd.DataFrame
         )
     ]
     return exploded
+
+
+################################################################################
+# XBRL Calculation Tree
+################################################################################
+Ferc1XbrlCalculationComponent = ForwardRef("Ferc1XbrlCalculationComponent")
+Ferc1XbrlCalculation = ForwardRef("Ferc1XbrlCalculation")
+
+
+class Ferc1XbrlCalculationComponent(BaseModel):
+    """A class representing the components of a FERC 1 XBRL Metadata calculation."""
+
+    name: str
+    weight: float
+    source_tables: list[str]
+    calculation: Ferc1XbrlCalculation | None = None
+
+    @validator("source_tables")
+    def has_unique_source_table(cls, value):
+        """Ensure that each calculation component has a unique source table."""
+        if len(value) == 1:
+            return value
+        else:
+            raise ValueError(
+                "Exploding tables requires calculations to have unique source tables."
+            )
+
+
+class Ferc1XbrlCalculation(BaseModel):
+    """A class representing a calculation from the FERC 1 XBRL Metadata."""
+
+    components: list[Ferc1XbrlCalculationComponent] = []
+    source_table: str
+    xbrl_factoid: str
+    xbrl_factoid_original: str
+
+    @property
+    def referenced_tables(self: Self) -> set[str]:
+        """Identify the set of all referenced tables in a calculation."""
+        return {c for x in self.calculations for c in x.source_tables}
+
+    def is_inter_table(self: Self, source_table: str | None = None) -> bool:
+        """Determine whether a calculation involves more than one table."""
+        all_tables = self.referenced_tables
+        if source_table is not None:
+            all_tables = all_tables.union({source_table})
+        return True if len(all_tables) > 1 else False
+
+
+Ferc1XbrlCalculationComponent.update_forward_refs()
+Ferc1XbrlCalculation.update_forward_refs()
+
+
+def resolve_calculation_components(
+    exploded_meta: pd.DataFrame,
+    table_name: str,
+    xbrl_factoid: str,
+) -> Ferc1XbrlCalculation:
+    """Build a calculation given exploded metadata, xbrl_factoid, and table_name."""
```

Comment:
Broadly, this recursive function looks fine, if I'm reading it right:

* base case is "If the exploded_meta for this table_name/xbrl_factoid has an empty list for "calculations", just return a `Ferc1XbrlCalculationComponent` with `components=[]`"
* otherwise, call this on whatever components this factoid points at

Though I suppose this assumes that walking the tree will eventually get you to nodes that have empty calculations.

I don't suppose there's a risk of hitting the recursion limit here before we get to the bottom of the tree?

Reply:
I don't think there's any very deep nesting -- maybe like 5-6 layers max. I think the tree I created in the comments on this issue is probably one of the biggest ones there is. And it's definitely our expectation that **eventually** all the components are reported values.  If we end up with cycles then either there's a problem with the specified calculations, or we've fundamentally misunderstood something.

There's another kind of termination condition that I don't think we've really conveyed here though, which is what happens if there's a reference to a table that is NOT in the set of tables that make up the explosion? In that case we also don't want continue recursing, and instead just refer to the reported value.  I think there was another edge-of-the-calculation-space case that I'm not remembering now though...